source/sigv4.c

```diff
@@ -793,6 +854,236 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+        if( sigV4Status == SigV4Success )
+        {
+            canonicalRequest->pBufCur--;
+            *canonicalRequest->pBufCur = '\n';
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t writeCanonicalHeaderToString( size_t headerIndex,
+                                                       CanonicalContext_t * canonicalRequest )
+    {
+        size_t buffRemaining, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0;
+        char * pBufLoc;
+        const char * value;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+        valLen = canonicalRequest->pHeadersLoc[ headerNum ].value.dataLen;
+
+        if( keyLen + valLen + 2 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
+                pBufLoc++;
+            }
+
+            *pBufLoc = ':';
+            pBufLoc++;
+            value = canonicalRequest->pHeadersLoc[ headerNum ].value.pData;
+            trimValueLen = 0;
+
+            for( i = 0; i < valLen; i++ )
+            {
+                if( !( isspace( value[ i ] ) && ( i + 1 <= valLen ) && ( ( i + 1 == valLen ) || isspace( value[ i + 1 ] ) || ( trimValueLen == 0 ) ) ) )
+                {
+                    *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].value.pData[ i ];
+                    pBufLoc++;
+                    trimValueLen + 1U;
+                }
+            }
+
+            *pBufLoc = '\n';
+            pBufLoc++;
+
+            buffRemaining -= ( keyLen + trimValueLen + 2 );
+
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = buffRemaining;
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendCanonicalizedHeaders( size_t headerCount,
+                                                     CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL );
+            sigV4Status = writeCanonicalHeaderToString( noOfHeaders, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t generateCanonicalHeaders( const char * pHeaders,
```

Comment:
This function is creating "CanonicalRequest" instead of "CanonicalHeaders"...
```suggestion
    static SigV4Status_t generateCanonicalRequest( const char * pHeaders,
```